src/providers/mod.rs

```diff
@@ -26,32 +26,47 @@ pub mod mbed_provider;
 #[cfg(feature = "tpm-provider")]
 pub mod tpm_provider;
 
-#[derive(Copy, Clone, Deserialize, Debug)]
-pub enum ProviderType {
-    MbedProvider,
-    Pkcs11Provider,
-    TpmProvider,
+#[derive(Deserialize, Debug)]
+#[serde(tag = "provider_type")]
+pub enum ProviderConfig {
+    MbedProvider {
+        key_id_manager: String,
+    },
+    Pkcs11Provider {
+        key_id_manager: String,
+        library_path: String,
+        slot_number: usize,
+        user_pin: Option<String>,
+    },
+    TpmProvider {
+        key_id_manager: String,
+        tcti: String,
+        owner_hierarchy_auth: String,
+    },
 }
 
-impl ProviderType {
-    pub fn to_provider_id(self) -> ProviderID {
-        match self {
-            ProviderType::MbedProvider => ProviderID::MbedProvider,
-            ProviderType::Pkcs11Provider => ProviderID::Pkcs11Provider,
-            ProviderType::TpmProvider => ProviderID::TpmProvider,
+use self::ProviderConfig::*;
```

Comment:
Just as an FYI, really - we tend not to do wildcard exports (unless it's for tests), either for enum variants or for module contents. We had to do that for some stuff around the TPM because there were so many things that we had to import... Generally in Rust world it's seen as an anti-pattern because it pollutes the namespace and makes it less obvious where identifiers are coming from.

For example, this is a lint that was introduced specifically for this case (glob import of enum variants) in clippy: https://github.com/rust-lang/rust-clippy/issues/541 - it is only a pedantic lint, though, which we don't enable (for now?)

Reply:
fixed